transport_nantes/transport_nantes/settings.py

```diff
@@ -102,11 +102,15 @@
 ]
 PASSWORD_RESET_TIMEOUT_DAYS = 1
 
-EMAIL_BACKEND = 'django_amazon_ses.EmailBackend'
-AWS_ACCESS_KEY_ID = settings_local.AWS_ACCESS_KEY_ID
-AWS_SECRET_ACCESS_KEY = settings_local.AWS_SECRET_ACCESS_KEY
-AWS_DEFAULT_REGION = settings_local.AWS_DEFAULT_REGION
-DEFAULT_FROM_EMAIL = settings_local.DEFAULT_FROM_EMAIL
+if ROLE == "dev":
```

Comment:
This is quite questionable and so needs extensive comment.  What exactly depends on this change?  Why did something local (like testing auth) require accommodation at global scope?

Is it not possible to modify the settings you need in the tests setup function?

Reply:
I've done that because Amazon SES doesn't work on local environment and makes this part bug with normal use and tests.
Adding this line also permit me to see sent mail and login/ reset link and that variables are well passed into the mail.
It is pretty usefull to see the mail and reset password sends mail. And like we don't want to change django librairie, I add to find a way to get the mail.

Reply:
Distinguish between your ad hoc testing during development (important) and unit tests, regression tests, etc. that are run automatically.  What you "see" is useful for you during development.  It is not related to automatic tests.

Why can't you change the settings value just in your test?

Reply:
I probably can do that for tests but website still doesn't run properly on usual use if I don't do that

Reply:
This assertion needs more detail.  "The website doesn't run" - what errors, what context, how, what do you expect, what do you observe?

Reply:
Here is the error from the website when it wants to send mail with Amazon SES:
>ClientError at /password_reset/
>An error occurred (InvalidClientTokenId) when calling the SendRawEmail operation: The security token included in the request is invalid.

Reply:
Yes, you don't have AWS SES credentials in dev.

Your job as a developer is not just to find bugs but to understand them, at least when it comes to the code base on which you're working, and ideally to propose fixes.

So, what's the stack trace?  Why is it trying to send mail with AWS?  How can we make it not try to send mail?  Because no mails should be sent out of dev, it's far too dangerous.

Reply:
Here is the error from the website when it wants to send mail with Amazon SES:
>ClientError at /password_reset/
>An error occurred (InvalidClientTokenId) when calling the SendRawEmail operation: The security token included in the request is invalid.

Reply:
1. It tries to send mail thru AWS because EMAIL_BACKEND constant in settings is set to use AWS.
2. password_reset uses PasswordResetForm which call send_mail function. So to not send mail, I would have to change this part of the code but I don't think that it's secure to modify django librairie. I may break the librairie or create a bug and mostly, we may have to do the job everytime django is upgraded. I'm pretty sure that nobody will think to write the changes again at each upgrade.
3. Send mail out of dev is too dangerous, so I must not use AWS. And I don't have credentials in dev so it is ok. But django allows me two others options:
  1. I can set django to write mail in file on my own computer in dev but I don't want to have plenty of useless files.
  2. I can set django print mail in terminal.
In both ways, mails stay on my machine, so in local. I prefer the second solution because the mail is emetted from local server to the terminal print every messages about the server itself. It stays on my computer, there is no save and I can easily use link inside the mail if needed.

Reply:
What does the code look like that tries to send mail?  What is the stack trace that gets there?  Can you see anything in that code or in the comments that indicates the design intentions of the authors?  Do you suppose the authors expected tests to send mails via production or other mail servers?

When tests pass, they should print almost nothing.  The state of passing or failing mustn't require a human to know anything about the test, only that it will say passed or failed.

Reply:
Here is the code from auth.views
`class PasswordResetView(PasswordContextMixin, FormView):
    email_template_name = 'registration/password_reset_email.html'
    extra_email_context = None
    form_class = PasswordResetForm
    from_email = None
    html_email_template_name = None
    subject_template_name = 'registration/password_reset_subject.txt'
    success_url = reverse_lazy('password_reset_done')
    template_name = 'registration/password_reset_form.html'
    title = _('Password reset')
    token_generator = default_token_generator

    @method_decorator(csrf_protect)
    def dispatch(self, *args, **kwargs):
        return super().dispatch(*args, **kwargs)

    def form_valid(self, form):
        opts = {
            'use_https': self.request.is_secure(),
            'token_generator': self.token_generator,
            'from_email': self.from_email,
            'email_template_name': self.email_template_name,
            'subject_template_name': self.subject_template_name,
            'request': self.request,
            'html_email_template_name': self.html_email_template_name,
            'extra_email_context': self.extra_email_context,
        }
        form.save(**opts)
        return super().form_valid(form)`

And the form used by above class:
`class PasswordResetForm(forms.Form):
    email = forms.EmailField(
        label=_("Email"),
        max_length=254,
        widget=forms.EmailInput(attrs={'autocomplete': 'email'})
    )

    def send_mail(self, subject_template_name, email_template_name,
                  context, from_email, to_email, html_email_template_name=None):
        """
        Send a django.core.mail.EmailMultiAlternatives to `to_email`.
        """
        subject = loader.render_to_string(subject_template_name, context)
        # Email subject *must not* contain newlines
        subject = ''.join(subject.splitlines())
        body = loader.render_to_string(email_template_name, context)

        email_message = EmailMultiAlternatives(subject, body, from_email, [to_email])
        if html_email_template_name is not None:
            html_email = loader.render_to_string(html_email_template_name, context)
            email_message.attach_alternative(html_email, 'text/html')

        email_message.send()

    def get_users(self, email):
        """Given an email, return matching user(s) who should receive a reset.
        This allows subclasses to more easily customize the default policies
        that prevent inactive users and users with unusable passwords from
        resetting their password.
        """
        email_field_name = UserModel.get_email_field_name()
        active_users = UserModel._default_manager.filter(**{
            '%s__iexact' % email_field_name: email,
            'is_active': True,
        })
        return (
            u for u in active_users
            if u.has_usable_password() and
            _unicode_ci_compare(email, getattr(u, email_field_name))
        )

    def save(self, domain_override=None,
             subject_template_name='registration/password_reset_subject.txt',
             email_template_name='registration/password_reset_email.html',
             use_https=False, token_generator=default_token_generator,
             from_email=None, request=None, html_email_template_name=None,
             extra_email_context=None):
        """
        Generate a one-use only link for resetting password and send it to the
        user.
        """
        email = self.cleaned_data["email"]
        if not domain_override:
            current_site = get_current_site(request)
            site_name = current_site.name
            domain = current_site.domain
        else:
            site_name = domain = domain_override
        email_field_name = UserModel.get_email_field_name()
        for user in self.get_users(email):
            user_email = getattr(user, email_field_name)
            context = {
                'email': user_email,
                'domain': domain,
                'site_name': site_name,
                'uid': urlsafe_base64_encode(force_bytes(user.pk)),
                'user': user,
                'token': token_generator.make_token(user),
                'protocol': 'https' if use_https else 'http',
                **(extra_email_context or {}),
            }
            self.send_mail(
                subject_template_name, email_template_name, context, from_email,
                user_email, html_email_template_name=html_email_template_name,
            )`
We can see that views 's class calls `form.save()` which sends mail.
Actually, I think that the authors have choosen to send mail for password recovery because of security purposes.
It is more secure than changing password directly on website. It is a kind of two-factors authentication but by mail instead of using phones.

Anyway, my problem comes from `send_mail()`. No matters from where it is called. settings is configured to use AWS right now. And for dev purposes, django has this solution:
`EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'`
and [I join the doc](https://docs.djangoproject.com/fr/2.2/topics/email/#console-backend) for full explaination.
Just one thing about that is that we have to remove this line for production

Reply:
I have that source code, too.

If you believe that the library requires sending mail (which is inappropriate in a test), that would be a bug.  Why don't you file a bug against the library?  Or ask in some discussion forum where the library maintainers hang out what they intended.  I'm sure you'll learn something...

If you believe what you are saying, act on it.
If you don't believe it, then do the work to understand the code.
Developing software is work.  It happens in the cool glow of a monitor, mostly alone.  Sometimes we ask other people for help.  Those people will expect to offer small bits of help, just to help us get unstuck.  Not to solve our problems.

Reply:
For testing, I can use the scratch code you gave me. It will work. And I don't really need mails in test so set `settings.ROLE = "test"` and change `if ROLE == "dev"` to `if ROLE == "test"` is perfect for the testing part to make send_mail() not to bug during tests. And for any values transmitted by mail, test code automatically generate them to simulate a sent mail. So for this part, it is ok

But where I am stuck is when I navigate on website in browser. It is not most of my work but I need to watch the design of the templates I write. 
Actually, I don't even need the mail, I need the token wrote in the mail to print pages.
It's about only 2 pages. So, for `send_activation()` you solved the problem. It still is password_reset template.
So, I got a solution from lots of different forums and even from django official documentation but you don't agree with it.
Now, I really don't know what to do except asking AWS credentials but it means sending mails on real mail account on web and we don't want it. It's dangerous for security of the project, it is spamming and most of all, we don't have unlimited mail address.